### ISSUE TITLE
Support for better game state visualizations

### ISSUE THREAD

**Issue by uduse** [deepmind/open_spiel#920]
I have been using OpenSpiel for my project and I found the string representation of the states are helpful but do have some limitations. I have been building visualization tools to use with OpenSpiel and I think it would be a nice contribution for OpenSpiel users, either as a part of OpenSpiel, or a standalone project.

### Render an image based on observation tensor
(Here, observation tensor means `np.array(state.observation_tensor()).reshape(game.observation_tensor_shape())`, an image in `HWC` format.)

In #678 I described the need of rendering human-readable game states from observation tensors, and rendering them as images is more human-friendly than ever (their visual representations were designed for human eyes at the first place). This way, the training pipeline does't have to store OpenSpiel game states. Instead, it can store observations as tensors, put them in a replay buffer and train the network with them, but still being human-readable. 

### Render image differences and continuations
I found tracing the changes between two string representations very difficult. There are two solutions that I found effective. One is to render trajectories (a stacked observation tensor in `BHWC` format) as GIFs, and the other one is to render the pixel differences between two rendered images of two consecutive states. I found using arrow keys to play the GIF back and forth gives me a much better idea of what's actually happening in the game.

## Example
My project (related #595) currently works well on Breakthrough, so I use it here as an example. This image is rendered based on the observation tensor of OpenSpiel's game state, the faint pink color indicates the difference from last state image.
![image](https://user-images.githubusercontent.com/4717005/187019481-b551bca3-2992-420e-aa03-bbe32d055879.png)

I also used these images to render the search tree using learned model (MuZero), and I found analyzing search trees with visually rich representation is easier than string representation.
![tempsnip](https://user-images.githubusercontent.com/4717005/187019261-35936dfc-deac-4a9a-90bd-17f8594efc92.png)

**Comment by rezunli96** [comment]
Hi @uduse, yeah good point! I think yes the current string or tensor representation of a state was not designed for visualization purposes in every game. From my experience in many games the tensor representations were designed to be providing the most accurate information of a state on the "machine-level", so you may expect lots of one-hot encodings in it. And using strings to represent states has been a common practice for people researching extensive-form games.
But I think they still provide sufficient information for one to write a customized function for visualization. In fact, I believe different people have different angles towards what is the "best visualization" of a state, and that is essentially a work of feature engineering. OpenSpiel kind of lefts it open to users to define their own way:)

**Comment by uduse** [comment]
@rezunli96 I agree! This totally sounds like a matter of taste. That's why I think maybe it's better for me to build the tool in a standalone project (and be unofficial) instead of being a part of OpenSpiel 🤔

**Comment by lanctot** [comment]
Hi @uduse, those visualizations look great!

I've always slightly preferred mapping from the state object themselves, because those observation tensors could change at any time (they tend to not, though!) and can break the visualization if so, whereas the state APIs will never change. Also if you do from observations or information sets, you can't visualize the imperfect information games. But I understand the more general thing is just not possible in your use case.

What GUI do you use to do the visualization?

Whether it'd make sense to include in OpenSpiel itself or separately depends on a few things. How much code would it be, and could it be nicely self-contained in its own directory? Most importantly: would you help maintain it if it breaks? If you put this up separately somewhere, I would happily link to it from the docs (as we've done with SpielViz).

**Comment by uduse** [comment]
> you can't visualize the imperfect information games

Yes, my approach gives up showing hidden information from the state and only visualize things that's also visible from the agent's point of view.

> What GUI do you use to do the visualization?

The GUI I use is any image viewer. I always work on a server and anything related to rendering to screen is more complicated. I found saving images and GIFs and opening them with VSCode has the less hassle among the solutions. I am not sure how many people share the same concern, maybe it's just me 🤣 

>  How much code would it be, and could it be nicely self-contained in its own directory? 
I would estimate 25 - 100 lines per game with some image assets, and it would be self-contained. My MuZero-based system works pretty well now and I will apply them to all two-players zero-sum perfect-information games in OpenSpiel. This means the project will at least include those games.

> Most importantly: would you help maintain it if it breaks?
That's a good question 😅  I can only guarantee short-term commitment and I would say that's the main reason I should do it in another repository. I will go ahead and do that.
